### ISSUE TITLE
standing page is listing a comment with zero flags

### ISSUE THREAD

**Issue by pushcx** [lobsters/lobsters#907]
A user reported this morning that the page at `https://lobste.rs/u/username/standing` listed a comment of theirs but not a list of flags for it, even after enough time had passed. Looking at it in the database, it has a `score` of `1` and `flags` of `0`. Looking in the `votes` table, the only vote for it is the upvote automatically created for the creator on posting.

The query in `UsersController#standing` for these (`flags > 0`) is simple, so I'm puzzled what's wrong and wanted to file this before I go to work for the day and risk forgetting about it. And, as always, I'm happy to have help diagnosing and fixing this, if anyone has spare attention before I get to it.

There's also a cosmetic bug, the vote reason is being displayed to mods as just `| `.

**Comment by threeplanetssoftware** [comment]
I started poking at this and I'm not sure if I've found a feature or a bug. If a bug, I'm not sure if it is a use case you are concerned about, or not. 

I may fall into the same camp as the user who reported this as right now one of my posts is showing as flagged, but there is no reason identified on the standing page (could be a legitimate flag, although the comment doesn't appear to fit any of the existing criteria) and it still has a score of 1. If you could check the production database for comment `mxwkbr`, that may help confirm.

Steps to reproduce what I'm finding on the test server:

1. Add a comment
2. Flag your own comment
3. Unflag own comment
   1. Check log for 500 server error in `update_score_and_recalculate!`

In order to test in a place I can see more, I am using the `fake_data` script on a current copy of the lobsters repo (none of my other PRs are applied, although the `fake_data` is from #916). I used the default `test` moderator account to comment on a random article and I can see the expected one vote:

```ruby
2.7.0 :037 > Vote.where(user_id: 2, story_id: 223, comment_id: 617)
  Vote Load (0.5ms)  SELECT `votes`.* FROM `votes` WHERE `votes`.`user_id` = 2 AND `votes`.`story_id` = 223 AND `votes`.`comment_id` = 617 LIMIT 11
 => #<ActiveRecord::Relation [#<Vote id: 848, user_id: 2, story_id: 223, comment_id: 617, vote: 1, reason: nil, updated_at: "2020-08-24 01:12:03">]> 
2.7.0 :038 > Comment.find(617)
  Comment Load (0.7ms)  SELECT `comments`.* FROM `comments` WHERE `comments`.`id` = 617 LIMIT 1
 => #<Comment id: 617, created_at: "2020-08-24 01:12:03", updated_at: "2020-08-24 01:12:03", short_id: "p5sny4", story_id: 223, user_id: 2, parent_comment_id: 616, thread_id: 457, comment: "Yerp", score: 1, flags: 0, confidence: 0.1828847834138887e0, markeddown_comment: "<p>Yerp</p>\n", is_deleted: false, is_moderated: false, is_from_email: false, hat_id: nil>
```

Once I click `flag` on myself, that changes to the below and I can visually see that it was marked off-topic on both the story and my standing page:

```ruby
2.7.0 :039 > Vote.where(user_id: 2, story_id: 223, comment_id: 617)
  Vote Load (0.5ms)  SELECT `votes`.* FROM `votes` WHERE `votes`.`user_id` = 2 AND `votes`.`story_id` = 223 AND `votes`.`comment_id` = 617 LIMIT 11
 => #<ActiveRecord::Relation [#<Vote id: 848, user_id: 2, story_id: 223, comment_id: 617, vote: -1, reason: "O", updated_at: "2020-08-24 01:22:09">]> 
2.7.0 :040 > Comment.find(617)
  Comment Load (0.4ms)  SELECT `comments`.* FROM `comments` WHERE `comments`.`id` = 617 LIMIT 1
 => #<Comment id: 617, created_at: "2020-08-24 01:12:03", updated_at: "2020-08-24 01:12:03", short_id: "p5sny4", story_id: 223, user_id: 2, parent_comment_id: 616, thread_id: 457, comment: "Yerp", score: -1, flags: 1, confidence: -0.19556271981131748e0, markeddown_comment: "<p>Yerp</p>\n", is_deleted: false, is_moderated: false, is_from_email: false, hat_id: nil> 
```

So far I believe this is expected behavior. However, when I click "unflag", things break. The "unflag" wording changes to "flag" as expected, but the comment is still marked as having 1 flag and has a score of -1. This is where I am unsure if I am seeing the same error that was reported, as you said in your instance `flags` explicitly was 0. 

```ruby
2.7.0 :041 > Vote.where(user_id: 2, story_id: 223, comment_id: 617)
  Vote Load (0.8ms)  SELECT `votes`.* FROM `votes` WHERE `votes`.`user_id` = 2 AND `votes`.`story_id` = 223 AND `votes`.`comment_id` = 617 LIMIT 11
 => #<ActiveRecord::Relation []> 
2.7.0 :042 > Comment.find(617)
  Comment Load (0.4ms)  SELECT `comments`.* FROM `comments` WHERE `comments`.`id` = 617 LIMIT 1
 => #<Comment id: 617, created_at: "2020-08-24 01:12:03", updated_at: "2020-08-24 01:12:03", short_id: "p5sny4", story_id: 223, user_id: 2, parent_comment_id: 616, thread_id: 457, comment: "Yerp", score: -1, flags: 1, confidence: -0.19556271981131748e0, markeddown_comment: "<p>Yerp</p>\n", is_deleted: false, is_moderated: false, is_from_email: false, hat_id: nil> 
```

The rails server console has this relevant output:

```ruby
Started POST "/comments/p5sny4/unvote" for 127.0.0.1 at 2020-08-23 21:24:00 -0400
Processing by CommentsController#unvote as */*
  Parameters: {"reason"=>"", "id"=>"p5sny4"}
  User Load (0.6ms)  SELECT `users`.* FROM `users` WHERE `users`.`session_token` = '1ZrokrektFgzqCdK4qpFWcjE9sLdy8ELknWMchr4H3z2EgFN7meFGTH59vpg' LIMIT 1
  ↳ app/controllers/application_controller.rb:30:in `authenticate_user'
  Request 127.0.0.1 POST /comments/p5sny4/unvote user: 2 test
  Keystore Load (0.2ms)  SELECT `keystores`.* FROM `keystores` WHERE `keystores`.`key` = 'traffic:at:5327440' LIMIT 1
  ↳ app/models/keystore.rb:13:in `value_for'
  Comment Load (0.3ms)  SELECT `comments`.* FROM `comments` WHERE `comments`.`short_id` = 'p5sny4' ORDER BY `comments`.`id` ASC LIMIT 1
  ↳ app/controllers/comments_controller.rb:371:in `find_comment'
  Vote Load (0.4ms)  SELECT `votes`.* FROM `votes` WHERE `votes`.`user_id` = 2 AND `votes`.`story_id` = 223 AND `votes`.`comment_id` = 617 ORDER BY `votes`.`id` ASC LIMIT 1
  ↳ app/controllers/comments_controller.rb:374:in `find_comment'
  CACHE Vote Load (0.0ms)  SELECT `votes`.* FROM `votes` WHERE `votes`.`user_id` = 2 AND `votes`.`story_id` = 223 AND `votes`.`comment_id` = 617 ORDER BY `votes`.`id` ASC LIMIT 1
  ↳ app/models/vote.rb:96:in `vote_thusly_on_story_or_comment_for_user_because'
   (0.1ms)  BEGIN
  ↳ app/models/vote.rb:113:in `vote_thusly_on_story_or_comment_for_user_because'
  Vote Destroy (0.3ms)  DELETE FROM `votes` WHERE `votes`.`id` = 848
  ↳ app/models/vote.rb:113:in `vote_thusly_on_story_or_comment_for_user_because'
   (0.9ms)  COMMIT
  ↳ app/models/vote.rb:113:in `vote_thusly_on_story_or_comment_for_user_because'
  Comment Load (0.3ms)  SELECT `comments`.* FROM `comments` WHERE `comments`.`id` = 617 LIMIT 1
  ↳ app/models/vote.rb:131:in `target'
/home/testbed/.rvm/gems/ruby-2.7.0/gems/transaction_isolation-1.0.5/lib/transaction_isolation/active_record/connection_adapters/mysql2_adapter.rb:59: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/home/testbed/.rvm/gems/ruby-2.7.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:609: warning: The called method `translate_exception_without_transaction_isolation_conflict' is defined here
   (0.7ms)  UPDATE comments SET
  score = (select sum(vote) from votes where comment_id = comments.id),
  flags = (select count(*) from votes where comment_id = comments.id and vote = -1),
  confidence = 0
WHERE id = 617

  ↳ app/models/comment.rb:319:in `update_score_and_recalculate!'
Completed 500 Internal Server Error in 20ms (ActiveRecord: 3.9ms | Allocations: 7613)

```
This then leads to the behavior it appears is described in the initial report (note the top comment is the one we messed with above, the bottom is one that is actually flagged). 
![image](https://user-images.githubusercontent.com/10688754/91015211-c9c38a80-e5b8-11ea-8f80-57e7fff5f2fc.png)

What appears to be happening is funkiness in the `vote_thusly_on_story_or_comment_for_user_because` method in `app/models/vote.rb`. If the new score coming in is 0, it deletes the vote. Because I am unvoting my -1 score, that means the incoming score is 0, so the vote will be deleted. Later in the method, it attempts to update the score and recalculate the flag change, but because the vote is already destroyed, it can't update it and errors out. Because it errors out, it never gets around to updating the flag count on the comment itself. 

```ruby
  def self.vote_thusly_on_story_or_comment_for_user_because(
    new_vote, story_id, comment_id, user_id, reason, update_counters = true
  )
    v = Vote.where(:user_id => user_id, :story_id => story_id,
      :comment_id => comment_id).first_or_initialize

    [snip]

    if new_vote == 0
      v.destroy!

    [snip]

    if update_counters
      [snip]
      t.update_score_and_recalculate!(score_delta, flag_delta)
    end
  end

```

Where I'm uncertain if this is a feature or bug is I'm not sure your assumptions on users flagging themselves and then unflagging. Seems unlikely, but it also seems to generate exactly the conditions you are describing. Where I'm not sure if this is the same bug, is your stated use case eexplicitly says that `flags` is still 0. That might mean it is another flavor of this bug that is different from the reproduction, or something else entirely.

Apologies for the novel, grateful for direction as to whether you think this is related based on what you can see in the production database and if this is a case worth catching, or just a feature of people that flag themselves in shame.

**Comment by pushcx** [comment]
@threeplanetssoftware and I spent some time kicking this around and he found the smoking gun, that new comments from fake_data showed up on the standing page. I realized it must be a default and fixed this in 5d89d5f8d37db52bc801fb92be9c9d0e699fa9c3.
